### ISSUE TITLE
esp32: need help: custom build is slower than stock from website

### ISSUE THREAD

**Issue by petrkr** [micropython/micropython#8029]
I trying speed up some TFT library and due LAN and ESPNow I still using own builds with SDK 3.3.2 (which is latest 1.14 upy). But I decided try first latest vanila uPy from website (night build) to compare if things speeds up meanwhile and yes they are.

But since I need do custom build due PSRAM and later with some pre-loaded libs etc I compile myself. But my custom build is in runtime slower than stock from night build.

I do not know why, so I want ask for help what can I have different in build environment. For now I use everything stock (git checkout, sdkcheckout, installed some binary toolchains by esp.idf 'install.sh' script)

Using SDK esp-idf tag 4.3 (not 4.3.1)
Using latest microypthon master (in time of writting it's 81f706aee

benchmark of TFT draw some box:

Custom build 1.14 with 3.3.2 SDK
 - 180ms (do not know what was freq now, but default)

official build 1.17 night 81f706aee (https://micropython.org/download/esp32/)
 - 104ms @ 160MHz
 - 90m @ 240MHz
 
custom build 1.17 81f706aee with 4.3 SDK
 - 140ms @ 160MHz
 - 110ms @ 240MHz
 
Would be nice to know to know used optimizations. For now I using board "GENERIC"

**Comment by petrkr** [comment]
Another test

custom build 1.17 81f706a with 4.5-dev SDK w/o SPI RAM
90ms @ 160MHz
70ms @ 240MHz

custom build 1.17 81f706a with 4.5-dev SDK with SPI RAM
115ms @ 160MHz
95ms @ 240MHz

**Comment by petrkr** [comment]
Another test:
seems official builds are from SDK 4.2.2 (or simillar 4.2.x)

custom build 1.17 81f706a with 4.2.2 SDK w/o SPI RAM
103ms @ 160MHz
85ms @ 240MHz

custom build 1.17 81f706a with 4.2.2 SDK with SPI RAM
150ms @ 160MHz
130ms @ 240MHz


Also free mem 109kB @ 4.2.2

others 4.3 and 4.5 only mem 62kB free, but I saw already issue about that #7962

**Comment by agatti** [comment]
It's been a while so probably it's not relevant anymore, I guess.  Prime candidate for proposed-close if nothing happens in a while.

**Comment by petrkr** [comment]
> It's been a while so probably it's not relevant anymore, I guess. Prime candidate for proposed-close if nothing happens in a while.

If it is solved, then I agree.

But after 5 years i really do not have that test cases.

Also if I remember well I used that for LN ATM and because of that I switched to DGUS UART LCD (something like Nextion).

But that does not mean 'fixed'

**Comment by agatti** [comment]
I tend to agree with you, however in the meantime the supported ESP-IDF version have changed, and with them the supported compiler versions.  Unless I'm mistaken the specific test case you used to reproduce the slowdown isn't available, so we only have MicroPython's own benchmarks.

I can do some tests on an ESP32 board tomorrow, but if you build your code with the same ESP-IDF + compiler (installed by ESP-IDF) version used by the build machine you should get the same binary except for maybe a git tag/timestamps mismatch.  As far as I know there's no special compilation flag that is used only for automated builds.

**Comment by dpgeorge** [comment]
The ESP32 can have very varied performance results due to its flash caching.  Sometimes a small change to the C code leads to a different layout of the code along different cache lines, and this changes the performance drastically.

We tried many things to stabilise the performance, and it's not too bad these days.  But in general this is a limitation of the hardware and the best thing you can do is just recompile with different options until its fast enough for you.
